### ISSUE TITLE
Support CUDA graph breaks

### ISSUE THREAD

**Issue by bm-synth** [thu-ml/SageAttention#74]
Hi @jason-huang03 
(cc @akedia in #73 )

Just found out that if you do `torch.compile(model, mode='max-autotune')` , then the model with `scaled_dot_product_attention` is faster than the model with SAGE attention.

That is because, the model that uses `scaled_dot_product_attention` allows for compilation without CUDA graph breaks, but the model with SAGE cant be compiled with CUDA full graphs, leading to no performance increase when compiled. 

The workaround is this:

```
# Wrappers to allow sage attention to be compiled with CUDA full graphs (max-autotune, reduce-overhead)
@torch.library.custom_op("mylib::forward", mutates_args=())
def sage_attention(q: torch.Tensor, k: torch.Tensor, v: torch.Tensor, tensor_layout: str) -> torch.Tensor:
    return sa.sageattn(q, k, v, tensor_layout=tensor_layout)


@sage_attention.register_fake
def _(q, k, v, tensor_layout):
    return torch.empty_like(q)
```

and then you call this `sage_attention` function **instead** of the `sa.sageattn` directly in your model. This `FakeTensor` will allow the whole model to be compiled with `max-autotune`.  [Here is some information about FakeTensor](https://pytorch.org/tutorials/advanced/python_custom_ops.html#python-custom-ops-tutorial).

@jason-huang03 is this something you'd like to integrate in the code base?

**Comment by jason-huang03** [comment]
Yes, we are working on it to make SageAttention compatible with torch.compile.

**Comment by jason-huang03** [comment]
Thank you for your advice!

**Comment by bm-synth** [comment]
@jason-huang03 just noticed my suggestion is not returning the correct output. im trying to find out why, any idea?

**Comment by jason-huang03** [comment]
Perhaps it is the `empty_like`, which might give incorrect stride information. Can you try `torch.empty` instead?

**Comment by bm-synth** [comment]
still the same noisy output :( do you have an example that worked for you?

**Comment by jason-huang03** [comment]
Currently not. I have tested with `torch.compile` with non-cuda-graph mode and it works fine (wrapping the `sage_attn` with `@torch.compiler.disable`). I shall try to compile with cuda graph and inform you the result.

**Comment by jason-huang03** [comment]
Sure!

**Comment by bm-synth** [comment]
So far:
- No compilation, no fake tensor: works
- No compilation, with fake tensors: works
- Model compiled with torch compile  `default`, no fake tensor: wrong results
- Model compiled with torch compile  `default`, with fake tensors:  wrong results
- Model compiled with torch compile `default`, with `@torch.compiler.disable` around Sage: works
- Model compiled with torch compile `max-autotune-no-cudagraphs`, with `@torch.compiler.disable` around Sage:  fails
- Model compiled with torch compile `max-autotune`, with `@torch.compiler.disable` around Sage: works with low performance (similar to no CUDA graphs) and prints several CUDA graph break warnings:

```
~/.venv/lib/python3.11/site-packages/torch/cuda/graphs.py:84: UserWarning: The CUDA Graph is empty. This usually means that the graph was attempted to be captured on wrong device or stream. (Triggered internally at ../aten/src/ATen/cuda/CUDAGraph.cpp:208.)
```

**Comment by bm-synth** [comment]
@jason-huang03 success!

To make it work, i needed to specify `q`, `k`, and `v` as mutable parameters, and return as you suggested a fake tensor as `torch.empty(*q.shape, dtype=q.dtype, device=q.device)` instead of `torch.empty_like(q)` 🚀 

I updated the PR message with the correct code.

**Comment by jason-huang03** [comment]
@bm-synth 
Awesome! We will include this feature in version 2.0.1 which will be available in a day or two.

**Comment by jason-huang03** [comment]
Hey @bm-synth I still got blurry result😂
My code is as this:
```python
from sageattention import sageattn
import torch.nn.functional as F

@torch.library.custom_op("mylib::sa_wrapper", mutates_args={"q", "k", "v"}, device_types="cuda")
def sa_wrapper(q: torch.Tensor, k: torch.Tensor, v: torch.Tensor, is_causal: bool = False) -> torch.Tensor:
    return sageattn(q, k, v, is_causal=is_causal)

@sa_wrapper.register_fake
def _(q, k, v, is_causal=False):
    return torch.empty(*q.shape, device=q.device, dtype=q.dtype)

F.scaled_dot_product_attention = sa_wrapper
```
Is there anything wrong with my code?

**Comment by bm-synth** [comment]
would that blurring happen if you call `sa_wrapper` directly? Also try to remove default variable assignments such as `is_causal: bool = False`. do you want to have a call to see it interactively?

**Comment by bm-synth** [comment]
@jason-huang03 you're right, `max-autotune` will not yield graph breaks warnings but will return noisy outputs.

I've been doing some debugging.

I checked the number of calls and input sizes when calling the compiled version of `sage_attention` and it matches the non-compiled implementation.

I also looked at the [The Custom Operators Manual](https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU/edit?tab=t.0#heading=h.oirj3xktyzui), it suggested to call a function registered as `@torch.library.custom_op("mylib::sage_attention"` via `torch.ops.mylib.sage_attention(q, k, v , ...)`. I tried that, didn't work. So I ran the `torch.library.opcheck` function to check if there was any registration errors. There are no errors.

As a side note, i updated this issue text with that code. 

For sanity checking, can you double check that your implementations follows the sections "How to add CPU/CUDA/Backend implementations" in [The Custom Operators Manual](https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU/edit?tab=t.0#heading=h.oirj3xktyzui) and [Custom C++ and CUDA Operators](https://pytorch.org/tutorials/advanced/cpp_custom_ops.html)?

I will now try to compare the input and output tensor **values** of the compiled vs uncompiled version. Meanwhile, we should bring this with the CogVideoX example to the torch developers for further help.

**Comment by bm-synth** [comment]
> I will now try to compare the input and output tensor values of the compiled vs uncompiled version. 

ok i checked, it's the output of `sageattn_qk_int8_pv_fp8_cuda` that is different on the fake tensors use case:
- if you compile with `max-autotune` without fake tensors (and with graph break warnings), `max-autotune-no-cudagraphs` without fake tensors, or `max-autotune-no-cudagraphs` with fake tensor, the outputs are the same.
- if you compile `max-autotune` with fake tensor, then the input to `sageattn_qk_int8_pv_fp8_cuda` is the same but the output is not.

We should ask torch devs for help. For the time being, can we allow the code to compile with cuda graphs and ignore sage?

**Comment by jason-huang03** [comment]
Nice observation. What do you mean by "allow code to compile with cuda graphs and ignore sage"? do you mean using the triton kernel instead of the cuda kernel?

**Comment by bm-synth** [comment]
> "allow code to compile with cuda graphs and ignore sage"

i meant instead of trying to have a CUDA graph of the whole run, that includes the sage operation and yields bad results, we can try to have a CUDA graph before, then sage, then another CUDA graph. ie implicit declare a compiled graph before, a compiled graph after, and ignore compilation on sage.

>   do you mean using the triton kernel instead of the cuda kernel?

Triton kernels work well with `max-aututone` and the fake tensors (I am not sure here, i will test again later), but at the moment we only have the triton kernlel for fp16 quantization. I believe the visible speedups on sage (in my use case and CogVideoX) are from the `int8` quantization. Are you planning to write the `int8` sage on triton as well?

**Comment by asahni04** [comment]
what is the fastest compile setting for cuda kernels fp8_fp16pv one?

**Comment by zou3519** [comment]
@bm-synth I'm one of the maintainers for torch.compile. Do you have an end-to-end script for your experiment? The outputs being incorrect is very worrying to us.

**Comment by pikerbright** [comment]
@bm-synth @zou3519 #162 
fix sage attention issue with torch compile
```
def sageattention_forward(
    query: torch.Tensor,
    key: torch.Tensor,
    value: torch.Tensor,
    attn_mask: Optional[torch.Tensor] = None,
    dropout_p: float = 0.0,
    is_causal: bool = False,
    *,
    scale: Optional[float] = None,
) -> torch.Tensor:
    query = query.permute(0, 2, 1, 3)
    key = key.permute(0, 2, 1, 3)
    value = value.permute(0, 2, 1, 3)
    out = sageattn(
        query,
        key,
        value,
        attn_mask=attn_mask,
        dropout_p=dropout_p,
        is_causal=is_causal,
        scale=scale,
        tensor_layout='NHD'
    )
    return out.permute(0, 2, 1, 3)
```

**Comment by jason-huang03** [comment]
@zou3519 we suppose it is because sage attention does not specify the cuda stream when launching the kernel. Could that be a problem?

**Comment by zou3519** [comment]
Yes, that can be a problem if sage attention gets launched on a stream that is different from the current cudastream set by PyTorch for capture.

This should be easy to test, here's an example of how to get PyTorch's current stream and pass it to the launch parameters: https://github.com/pytorch/pytorch/blob/78684e27ace1ebd51ccc93d4b6b8a6a3ec3b4c14/aten/src/ATen/native/cuda/WeightNorm.cu#L365C18-L365C24

**Comment by woct0rdho** [comment]
As of PyTorch 2.9, `torch.cuda.get_device_capability` cannot be traced by torch.compile. So I made this change:
https://github.com/woct0rdho/SageAttention/commit/358a705c135c4565a153f092d9a9ad2e792a1f46

This along with the recent https://github.com/thu-ml/SageAttention/pull/218 allows me to compile `sageattn` with `fullgraph=True`, at least in the single-GPU case. We still need to check how `torch.cuda.set_device` affects the multi-GPU case.
